h5netcdf/core.py

```diff
@@ -234,6 +239,46 @@ def __init__(self, parent, name):
                     if k.startswith('_nc4_non_coord_'):
                         var_name = k[len('_nc4_non_coord_'):]
                     self._variables.add(var_name)
+
+        # One last pass to get the current sizes. This is necessary as h5netcdf
+        # needs to now the current size of each unlimited dimension.
+        #
+        # I'm not entirely sure how the netcdf C api determines the current
+        # size - should probably be looked into.
+        def _find_dim(h5group, dim):
+            if dim not in h5group:
+                # Might happen if the dimension has not been created yet.
+                if h5group.name == "/":
```

Comment:
I don't think this can happen here. We are only checking dimensions in `self.dimensions`, which looks like at `self._dim_sizes` -- which in turn is only filled in where there is already a matching dimension scale.

Reply:
With the lazy evaluation it can happen - all my changed lines (except the `#pragma: no cover` ones) are covered by the tests. But we can skip the whole dimension detection logic if the init actually created the file because then we are already internally consistent. I'll change it.

Reply:
Lazy evaluation controls creating sub-groups/child-variables in h5netcdf, but it shouldn't dictate what appears in `self._h5group` (which is based on what's in the file per h5py). So I'm somewhat mystified how this can happen.

Can you point to which test triggers this line (e.g., if you add a print statement)?

Reply:
It can no longer happen with my latest changes as the logic to determine the current sizes of the unlimited dimensions is now only called when reading from an existing file.

It might be that I misunderstood how the lazy evaluation actually works but the actual dimension scales in the HDF5 file are only created during the `flush()` call, so initializing a new file, adding a dimension, and then creating a new group would call the `__init__()` method with `self.dimensions` being set, without dimension scales being in the file (yet).

Reply:
OK, thanks for your patience here. You're right -- we have two forms of laziness here (and I had forgotten how my code worked!).